### PR TITLE
Added ExprArrowKnockbackStrength

### PR DESCRIPTION
**Pull request by DeltaRays**
---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET || mode == ChangeMode.RESET) ? CollectionUtils.array(Number.class) : null;
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		int state = delta != null ? ((Long) delta[0]).intValue() : 1;
```
Comment (Matocolotoe) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:60]: `((Long) delta[0]).intValue()`
Just cast the object to an integer ?

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:60]: You want to cast it to a Number, not a Long

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: It told me it couldn't cast a Long to an Integer

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: (to Matocolotoe)

File: src/main/java/ch/njol/skript/expressions/ExprProjectileKnockbackStrength.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET || mode == ChangeMode.RESET) ? CollectionUtils.array(Boolean.class) : null;
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		int strength = delta != null ? ((Long) delta[0]).intValue() : 1;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileKnockbackStrength.java:68]: You should cast to Number, not Long 🙂

Also, how are values < 0 handled? Do they work?

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET || mode == ChangeMode.RESET) ? CollectionUtils.array(Number.class) : null;
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		int strength = delta != null ? Math.max(((Number) delta[0]).intValue(), 0) : 1;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:67]: You might be better off checking if strength < 0 and just not doing anything, but that’s up for debate: should setting it to a negative value set it to no strength? Personally I don’t think it should do anything, but others may disagree.

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		for (Projectile entity : getExpr().getAll(e)) {
+			if (abstractArrowExists) {
+				System.out.println(entity.getType().name());
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:70]: Accidental debug line left in 😎

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:70]: oop

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java (hunk 1 above)
Comment (Matocolotoe) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: @DeltaRays I was talking about `delta[0]`

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: Yeah, so was I

Comment (Matocolotoe) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: ?????

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	"\tevent-projectile is an arrow",
+	"\tset arrow knockback strength of event-projectile to 10"})
+@Since("INSERT VERSION")
+public class ExprArrowKnockbackStrength extends SimplePropertyExpression<Projectile, Number> {
+	
+	static {
+		register(ExprArrowKnockbackStrength.class, Number.class, "[the] (arrow) knockback strength", "projectiles");
+	}
+	
+	boolean abstractArrowExists = Skript.classExists("org.bukkit.entity.AbstractArrow");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:49]: This should be a constant

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET || mode == ChangeMode.RESET) ? CollectionUtils.array(Number.class) : null;
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		int strength = delta != null ? Math.max(((Number) delta[0]).intValue(), 0) : 1;
+		for (Projectile entity : getExpr().getAll(e)) {
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:68]: Always use Expression#getArray over getAll

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+@Name("Arrow Knockback Strength")
+@Description("An arrow's knockback strength.")
+@Examples({"on shoot:",
+	"\tevent-projectile is an arrow",
+	"\tset arrow knockback strength of event-projectile to 10"})
+@Since("INSERT VERSION")
+public class ExprArrowKnockbackStrength extends SimplePropertyExpression<Projectile, Number> {
+	
+	static {
+		register(ExprArrowKnockbackStrength.class, Number.class, "[the] (arrow) knockback strength", "projectiles");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:46]: Why is `arrow` between parenthesis?

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Arrow Knockback Strength")
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:37]:
```suggestion
@Name("Projectile Knockback Strength")
```

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+@Description("An arrow's knockback strength.")
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:38]:
```suggestion
@Description("The knockback strength of projectiles considered arrows (i.e.: tridents as well).")
```

**Review by FranKusmiruk — DISMISSED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java (hunk 1 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: When I casted delta[0] to an integer it told me it couldn’t cast a long to an integer. Don’t worry, I’ve fixed it

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java (hunk 8 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:37]: F74F8B90-1A38-4A4C-A042-539F777A1DBF

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java (hunk 9 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:38]: 0369A32F-A9F8-4F4B-8AEF-3178E5BD9051

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java (hunk 1 above)
Comment (Matocolotoe) [reply src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java]: Of course but I'm talking about the original object... Nevermind.

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+			case RESET:
+			case SET:
+				for (Projectile entity : getExpr().getArray(e)) {
+					if (abstractArrowExists) {
+						if (entity instanceof AbstractArrow) ((AbstractArrow) entity).setKnockbackStrength(strength);
+					} else if (entity instanceof Arrow) {
+						((Arrow) entity).setKnockbackStrength(strength);
+					}
+				}
+				break;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:118]: Default case 😎

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	public Number convert(Projectile arrow) {
+		if (abstractArrowExists)
+			return arrow instanceof AbstractArrow ? ((AbstractArrow) arrow).getKnockbackStrength() : null;
+		return arrow instanceof Arrow ? ((Arrow) arrow).getKnockbackStrength() : null;
+	}
+	
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.DELETE || mode == ChangeMode.REMOVE_ALL) ? null : CollectionUtils.array(Number.class);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:62]: Maybe use a switch statement here?

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+				return null;
+		}
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		int strength = delta != null ? ((Number) delta[0]).intValue() : 0;
+		switch (mode) {
+			case REMOVE:
+				if (abstractArrowExists)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprArrowKnockbackStrength.java:78]: Maybe not the last thing after all haha. These if and else statements should have curly braces.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Review by ShaneBeee — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]

**Pull request merged**